optaplanner-website-root/content/blog/2022/06/09/RunOptaPlannerWorkloadOnOpenShift.adoc

```diff
@@ -0,0 +1,231 @@
+= Run OptaPlanner workloads on OpenShift, part I.
+rsynek
+2022-06-09
+:page-interpolate: true
+:jbake-type: post
+:jbake-tags: cloud, openshift, kubernetes
+
+After you have defined the problem domain model, implemented constraints and benchmarked the best configuration
+for your planning problem, you might be thinking that you are done. In fact, yet another fun is just about to begin.
+
+OptaPlanner does not exist in an isolated world of ready-to-be-executed datasets.
+You have to deploy the solver project to the target environment, these days likely some kind of cloud platform.
+Another question immediately pops up in one's mind: can OptaPlanner leverage the cloud platform and scale horizontally?
+
+Recently, we have added a new, experimental, `optaplanner-operator` module that will simplify running OptaPlanner workloads
+on Kubernetes.
+
+In this article I am going to show how to use the OptaPlanner Operator to deploy and scale a school timetabling solver on https://www.redhat.com/en/technologies/cloud-computing/openshift[OpenShift].
+
+The source code is available in the https://github.com/kiegroup/optaplanner-quickstarts/tree/development/technology/kubernetes[kubernetes demo],
+along with the https://github.com/kiegroup/optaplanner-quickstarts[optaplanner-quickstarts].
+
+== Architecture
+
+image::demoArchitecture.svg[OpenShift demo architecture diagram]
+
+The demo consists of two projects - the School Timetabling, which defines the optimization problem, and the Demo App,
+which generates datasets and sends requests to the School Timetabling. Both these projects are https://quarkus.io/[Quarkus] applications.
+
+These two parts communicate via Kafka topics created by the OptaPlanner Operator: the `school-timetabling-problem`
+and the `school-timetabling-solution`. The Demo App stores a dataset into the PostgreSQL database and sends a message
+to the `school-timetabling-problem` topic. The School Timetabling reads the message, loads the dataset from the database
+and solves it. After that, it stores the solution back to the database and sends a message to the `school-timetabling-solution`
+topic to let the Demo App know the solution is ready for taking.
+
+=== Solver Operator
+
+The OptaPlanner Operator is a Quarkus application developed on top of the https://javaoperatorsdk.io[javaoperatorsdk].
+Its job is to ensure all the Kubernetes resources the solver needs are in place: it creates Kafka topics and a deployment that
+runs the solver project, in this case, the School Timetabling.
+
+== Running the demo
```

Comment:
I am not entirely sure if we want to duplicate the READMEs here.
This is a highly detailed process and as OpenShift evolves, it will evolve with it.
I would rather just reference the READMEs here.

Reply:
Good point. In the end, I moved the entire setup to the demo and left just the scaling part of it here.